x/evm/abci_test.go

```diff
@@ -1039,6 +1042,102 @@ func TestHandleTransferKey(t *testing.T) {
+		ctx, bk, n, s, sourceCk, destinationCk = setup()
+		confirmedEventQueue = &utilsMock.KVQueueMock{}
+		bk.ForChainFunc = func(chain nexus.ChainName) types.ChainKeeper {
+			switch chain {
+			case sourceChainName:
+				return sourceCk
+			case destinationChainName:
+				return destinationCk
+			default:
+				return nil
+			}
+		}
+		bk.HasChainFunc = func(ctx sdk.Context, chain nexus.ChainName) bool { return true }
+
+		n.GetChainFunc = func(ctx sdk.Context, chain nexus.ChainName) (nexus.Chain, bool) {
+			switch chain {
+			case sourceChainName, destinationChainName:
+				return nexus.Chain{Name: chain}, true
+			default:
+				return nexus.Chain{}, true
+			}
+		}
+		n.IsChainActivatedFunc = func(ctx sdk.Context, chain nexus.Chain) bool { return true }
+		n.GetChainsFunc = func(_ sdk.Context) []nexus.Chain {
+			return []nexus.Chain{{Name: sourceChainName}}
+		}
+
+		destinationCk.GetChainIDFunc = func(ctx sdk.Context) (sdk.Int, bool) { return sdk.ZeroInt(), true }
+		destinationCk.EnqueueCommandFunc = func(ctx sdk.Context, cmd types.Command) error { return nil }
+		destinationCk.GetGatewayAddressFunc = func(sdk.Context) (common.Address, bool) {
+			return common.BytesToAddress(rand.Bytes(common.AddressLength)), true
+		}
+
+		s.GetCurrentKeyIDFunc = func(ctx sdk.Context, chain nexus.Chain, keyRole tss.KeyRole) (tss.KeyID, bool) {
+			return tssTestUtils.RandKeyID(), true
+		}
+		s.GetNextKeyIDFunc = func(ctx sdk.Context, chain nexus.Chain, keyRole tss.KeyRole) (tss.KeyID, bool) { return "", false }
+
+		sourceCk.SetEventCompletedFunc = func(ctx sdk.Context, eventID types.EventID) error { return nil }
+		sourceCk.GetConfirmedEventQueueFunc = func(sdk.Context) utils.KVQueue {
+			return confirmedEventQueue
+		}
+	})
+
+	givenConfirmedEventQueue.
+		When("events in queue exceeds limit", func() {
+			confirmedEventQueue.IsEmptyFunc = func() bool { return false }
+			event := types.Event{
+				Chain: sourceChainName,
+				TxId:  evmTestUtils.RandomHash(),
+				Index: uint64(rand.PosI64()),
+				Event: &types.Event_ContractCall{
+					ContractCall: &types.EventContractCall{
+						Sender:           evmTestUtils.RandomAddress(),
+						DestinationChain: destinationChainName,
+						ContractAddress:  evmTestUtils.RandomAddress().Hex(),
+						PayloadHash:      types.Hash(evmCrypto.Keccak256Hash(rand.Bytes(100))),
+					},
+				},
+			}
+			confirmedEventQueue.DequeueUntilFunc = func(value codec.ProtoMarshaler, filter func(value codec.ProtoMarshaler) bool) bool {
+				bz, _ := event.Marshal()
+				if err := value.Unmarshal(bz); err != nil {
+					panic(err)
+				}
+
+				return true
+			}
+		}).
+		Then("should handle limited number of events", func(t *testing.T) {
+			err := handleConfirmedEvents(ctx, bk, n, s)
+			assert.NoError(t, err)
+			assert.Len(t, sourceCk.SetEventCompletedCalls(), maxConfirmedEventsPerBlock)
+			assert.Len(t, destinationCk.EnqueueCommandCalls(), maxConfirmedEventsPerBlock)
```

Comment:
Maybe add a check to see if queue is empty after the max is reached? Can make `IsEmpty`/`DequeueUntil` return `False` after being called `x` times.